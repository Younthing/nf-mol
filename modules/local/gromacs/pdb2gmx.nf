process PDB2GMX_GROMACS {
    tag "${meta.id}"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://depot.galaxyproject.org/singularity/gromacs:2021.3--mpi_openmpi_h9969a6a_0'
        : 'biocontainers/gromacs:2021.3--mpi_openmpi_h9969a6a_0'}"

    input:
    tuple val(meta), path(pdb)

    output:
    tuple val(meta), path("*.gro"), emit: structure
    tuple val(meta), path("*.top"), emit: topology
    tuple val(meta), path("*.itp"), emit: include_topology
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    gmx pdb2gmx \
        -f ${pdb} \
        -o ${prefix}_processed.gro \
        -p ${prefix}_topol.top \
        -i ${prefix}_posre.itp \
        -water tip3p \
        -ff charmm27 \
        -ignh  # 忽略氢原子\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gromacs: \$(gmx --version | grep 'GROMACS version' | grep -oE '[0-9]+[0-9]+')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_processed.gro
    touch ${prefix}_topol.top
    touch ${prefix}_posre.itp

    cat <<-END_VERSIONS > versions.yml
    
    "${task.process}":
        gromacs: \$(gmx --version | gmx --version | grep 'GROMACS version' | grep -oE '[0-9]+[0-9]+')
    END_VERSIONS
    """
}
