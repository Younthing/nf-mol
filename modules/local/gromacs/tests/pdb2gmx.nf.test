nextflow_process {

    name "Test Process PDB2GMX_GROMACS"
    script "../pdb2gmx.nf"
    process "PDB2GMX_GROMACS"

    test("1fjs_protein[pdb]") {

        when {
            params {
                outdir = "tests/results"
            }
            process {
                """
                def meta = [id: '1fjs']
                input[0] = tuple(meta, file("${projectDir}/data/gromacs/1fjs_protein.pdb"))
                """
            }
        }

        then {
            assert process.success
            
            // 验证结构和拓扑文件存在
            assert path("${process.out.structure.get(0).get(1)}").exists()
            assert path("${process.out.topology.get(0).get(1)}").exists()
            
       
            // 验证文件名格式
            assert process.out.structure.get(0).get(1).toString().endsWith('.gro')
            assert process.out.topology.get(0).get(1).toString().endsWith('.top')
            
            // 验证主要文件非空
            assert path("${process.out.structure.get(0).get(1)}").size() > 0
            assert path("${process.out.topology.get(0).get(1)}").size() > 0
            
        }

    }

}
