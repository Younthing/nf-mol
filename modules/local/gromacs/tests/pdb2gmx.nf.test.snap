{
    "1fjs_protein[pdb]": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1fjs"
                        },
                        "1fjs_processed.gro:md5,439e4304437c22e60b8a4300069c2509"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "1fjs"
                        },
                        "1fjs_topol.top:md5,ddacd66c6b009af725772d77b44b0952"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "1fjs"
                        },
                        [
                            "1fjs_posre_Protein_chain_A.itp:md5,908b0308e85a68d038e77e005b53ec41",
                            "1fjs_posre_Protein_chain_L.itp:md5,cac419e91aed41e5e9eed6b0a4ba71d2",
                            "1fjs_topol_Protein_chain_A.itp:md5,ffb4d08c81851bb2b06db94a9e6afed5",
                            "1fjs_topol_Protein_chain_L.itp:md5,8e34006b01d70f6159fccd5d51de08e3"
                        ]
                    ]
                ],
                "3": [
                    "versions.yml:md5,a6297bdeaa47b4311c23eeb3f0ddfd40"
                ],
                "include_topology": [
                    [
                        {
                            "id": "1fjs"
                        },
                        [
                            "1fjs_posre_Protein_chain_A.itp:md5,908b0308e85a68d038e77e005b53ec41",
                            "1fjs_posre_Protein_chain_L.itp:md5,cac419e91aed41e5e9eed6b0a4ba71d2",
                            "1fjs_topol_Protein_chain_A.itp:md5,ffb4d08c81851bb2b06db94a9e6afed5",
                            "1fjs_topol_Protein_chain_L.itp:md5,8e34006b01d70f6159fccd5d51de08e3"
                        ]
                    ]
                ],
                "structure": [
                    [
                        {
                            "id": "1fjs"
                        },
                        "1fjs_processed.gro:md5,439e4304437c22e60b8a4300069c2509"
                    ]
                ],
                "topology": [
                    [
                        {
                            "id": "1fjs"
                        },
                        "1fjs_topol.top:md5,ddacd66c6b009af725772d77b44b0952"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a6297bdeaa47b4311c23eeb3f0ddfd40"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-16T17:56:53.276971962"
    }
}